test/integration/integration_test.cc

```diff
@@ -425,7 +425,8 @@ TEST_P(IntegrationTest, TestSmuggling) {
                                 "identity,chunked \r\ncontent-length: 36\r\n\r\n" +
                                 smuggled_request;
     sendRawHttpAndWaitForResponse(lookupPort("http"), request.c_str(), &response, false);
-    EXPECT_THAT(response, HasSubstr("HTTP/1.1 400 Bad Request\r\n"));
+    // 'identity' encoding is not supported
+    EXPECT_THAT(response, HasSubstr("HTTP/1.1 501 Not Implemented\r\n"));
```

Comment:
This seems not right. Why are we now returning a 501 here?

Reply:
onHeadersCompleteBase checks for unsupported encodings (`identity` in this case) and returns 501 if unsupported encoding is detected.
if everything is ok - passes control to onHeadersComplete() where my code is executed.

I can refactor it to do CL/TE checks before TE encoding check which returns 501, I'm not sure which order is correct in this case.

Reply:
nginx's response in case of unsupported encoding and CL set:
```
> POST / HTTP/1.1
> Host: example.com
> User-Agent: curl/7.64.1
> Accept: */*
> Transfer-Encoding: unsupported
> Content-Length: 10
>
< HTTP/1.1 501 Not Implemented
< Server: nginx
< Date: Fri, 31 Jul 2020 20:40:47 GMT
< Content-Type: text/html
< Content-Length: 14480
< Connection: close
```

Reply:
Naively it seems like we should return a 4xx code here, but I'm not an expert. If this is what NGINX does that seems fine, thanks.

Reply:
OK I see by allowing this in the parser we get slightly different behavior here. I think this is OK but I think we should release note this also and I would also like to hear from @antoniovicente or @PiotrSikora on this.

Reply:
Friendly ping on the release note on the subtle behavior change here.

/wait

Reply:
4xx makes more sense to me in this case.

This specific change seems unrelated to the rest of the functional changes in this PR, I would suggest doing the http_parser version bump separately from the rest of this PR.

Reply:
I can split this PR in two, one for library update and one for functional change.

Reply:
> This specific change seems unrelated to the rest of the functional changes in this PR, I would suggest doing the http_parser version bump separately from the rest of this PR.

@veshij correct me if I'm wrong, but I don't think this has to do with the library update. It has to do with the fact that we are now hitting code for this case that we did not hit before, and Envoy is returning 501 when it used to return 400. Right? If so I think this seems OK to me with a release note but will defer to @antoniovicente and @PiotrSikora on that.

Reply:
>I would suggest doing the http_parser version bump separately
Sorry, that was a reply to comment above.